src/x2/LteX2Manager.cc

```diff
@@ -96,66 +104,60 @@ void LteX2Manager::handleMessage(cMessage *msg)
     }
 }
 
-void LteX2Manager::fromStack(cPacket* pkt)
+void LteX2Manager::fromStack(Packet* pkt)
 {
-    LteX2Message* x2msg = check_and_cast<LteX2Message*>(pkt);
-    X2ControlInfo* x2Info = check_and_cast<X2ControlInfo*>(x2msg->removeControlInfo());
+    auto x2msg = pkt->peekAtFront<LteX2Message>();
+    auto x2Info = pkt->removeTagIfPresent<X2ControlInfoTag>();
 
     if (x2Info->getInit())
     {
         // gate initialization
         LteX2MessageType msgType = x2msg->getType();
-        int gateIndex = x2msg->getArrivalGate()->getIndex();
+        int gateIndex = pkt->getArrivalGate()->getIndex();
         dataInterfaceTable_[msgType] = gateIndex;
 
         delete x2Info;
-        delete x2msg;
+        delete pkt;
         return;
     }
 
+
     // If the message is a HandoverDataMsg, send to the GTPUserX2 module
-    if (x2msg->getType() == X2_HANDOVER_DATA_MSG)
-    {
-        // GTPUserX2 module will tunnel this datagram towards the target eNB
-        DestinationIdList destList = x2Info->getDestIdList();
-        DestinationIdList::iterator it = destList.begin();
-        for (; it != destList.end(); ++it)
-        {
-            X2NodeId targetEnb = *it;
-            x2msg->setSourceId(nodeId_);
-            x2msg->setDestinationId(targetEnb);
+    // (GTPUserX2 module will tunnel this datagram towards the target eNB)
+    // otherwise it is a X2 control message and sent to the x2 peer
 
-            // send to the gate connected to the GTPUser module
-            cGate* outputGate = gate("x2Gtp$o");
-            send(x2msg, outputGate);
-        }
-        delete x2Info;
-    }
-    else  // X2 control messages
+    DestinationIdList destList = x2Info->getDestIdList();
+    DestinationIdList::iterator it = destList.begin();
+
+    for (; it != destList.end(); ++it)
     {
-        DestinationIdList destList = x2Info->getDestIdList();
-        DestinationIdList::iterator it = destList.begin();
-        for (; it != destList.end(); ++it)
-        {
-            // send a X2 message to each destination eNodeB
-            LteX2Message* x2msg_dup = x2msg->dup();
-            x2msg_dup->setSourceId(nodeId_);
-            x2msg_dup->setDestinationId(*it);
+        X2NodeId targetEnb = *it;
+        auto pktDuplicate = pkt->dup();
+        auto updatedX2Msg = pktDuplicate->removeAtFront<LteX2Message>();
+        updatedX2Msg->markMutableIfExclusivelyOwned();
```

Comment:
The removed chunk is mutable by default, you don't have to call the markMutable... function.

Reply:
Call to markMutable was removed.